pipert/core/routine.py

```diff
@@ -153,6 +154,24 @@ def remove_event_handler(self, handler, event_name):
                              " event handlers".format(handler))
         self._event_handlers[event_name] = new_event_handlers
 
+    def pace(self, fps):
+        """Pace the routine to work at a wanted fps
+
+        Args:
+            fps: The wanted fps for the routine
+        """
+        def start_time(routine: Routine):
+            routine.state.start_time = time.time()
+
+        def start_pacing(routine: Routine, requested_fps):
+            if routine.state.output:
+                excess_time = (1 / requested_fps) - (time.time() - routine.state.start_time)
+                if excess_time > 0:
+                    time.sleep(excess_time)
+
+        self._event_handlers[Events.BEFORE_LOGIC].insert(0, (start_time, (), {}))
+        self._event_handlers[Events.AFTER_LOGIC].insert(0, (start_pacing, (fps,), {}))
```

Comment:
Maybe you should use the `add_event_handler` method instead of adding explicitly for readability sake.

Reply:
add_event_handler always appends the handler given to it and I want both of my timings to work instantly with the start of the routing and when it finishes so I want to guarantee that they would always be in the beginning of the list.

Maybe I should add a function very similar to add_event_handler?

Reply:
hmm, it raises 2 questions:

1. Do we want the pacing to be applied only to the `main logic` or we want to include the before/after logic handlers in the time measurement?
2. Do we want to add an optional functionality to `add_event_handler` to specify if when a handler should be called (first / last / doesn't matter)

What do you think? @Elon-Abulafia

Reply:
1. Maybe we should apply the time measurements on the whole component process, meaning: at the start of `before logic` and at the end of `after logic`, instead of applying it only to the `main logic`.

2. That would be a better design choice in my opinion.

@ItamarWilf

Reply:
Sounds good. Do you want to go ahead and implement it?

Reply:
Yeah i'll do that